src/main/kotlin/org/rust/ide/template/postfix/stringBasedPostfixTemplates.kt

```diff
@@ -0,0 +1,76 @@
+package org.rust.ide.template.postfix
+
+import com.intellij.codeInsight.template.Template
+import com.intellij.codeInsight.template.impl.TextExpression
+import com.intellij.codeInsight.template.postfix.templates.StringBasedPostfixTemplate
+import com.intellij.psi.PsiElement
+import org.rust.lang.core.psi.RustBinaryExprElement
+import org.rust.lang.core.psi.RustExprElement
+import org.rust.lang.core.psi.RustTupleFieldDeclElement
+import org.rust.lang.core.psi.impl.RustPsiImplUtil
+import org.rust.lang.core.psi.util.descendentsOfType
+import org.rust.lang.core.psi.util.parentOfType
+import org.rust.lang.core.resolve.RustResolveEngine
+import org.rust.lang.core.resolve.scope.RustResolveScope
+import org.rust.lang.core.types.RustEnumType
+import org.rust.lang.core.types.util.resolvedType
+
+class MatchPostfixTemplate : StringBasedPostfixTemplate(
+    "match",
+    "match expr {...}",
+    RustTopMostInScopeSelector(RustExprElement::isEnum)
+) {
+    override fun getTemplateString(element: PsiElement): String? {
+        val enumType = (element as RustExprElement).resolvedType as RustEnumType
+
+        val allAvailableNames = generateSequence(
+            element.parentOfType<RustResolveScope>(),
+            { it.parentOfType<RustResolveScope>() }
+        )
+            .flatMap { RustResolveEngine.declarations(it, element) }
+            .mapNotNull { it.element }
+            .mapNotNull { RustPsiImplUtil.canonicalCratePath(it) }
+            .map { it.toString() }
+            .toSet()
+
+        val stringBuilder = StringBuilder()
+        stringBuilder.append("match ${element.text} {\n")
+
+        for ((i, item) in enumType.item.enumBody.enumVariantList.withIndex()) {
+            val itemCanonicalName = RustPsiImplUtil.canonicalCratePath(item).toString()
+
+            val excessPath = allAvailableNames
```

Comment:
>This is a costly computation which can be hoisted out of the loop.

Sorry, of course it can not be moved out of the loop because it uses `itemCanonicalName`. 

The problem with correctness can arise due to aliases: 

```
mod m {
    pub enum E {
        X
    }
}

use self::m::E as S;

fn main() {
    match S::X {
        S::X => {}
    }
}
```

But yes, adding a prefix when there's an imported variant in scope seems bad. 


What do you think about the following rule: "either all variants should be qualified by enum name, or none should be qualified". 

That, is the following are possible expansions:

```
match expr {
  Expr::Add() => {}
  Expr::Sub() => {}
}

match expr {
  Add() => {}
  Sub() => {}
}
```

and mixing or further qualifying is not allowed

```
match expr {
  Expr::Add() => {}
  Sub() => {}
}
match expr {
  ast::Expr::Add() => {}
  ast::Expr::Sub() => {}
}
```

?

That way, I think you can lookup only the first variant in the scope, which should allow to avoid the creation of list or set of names altogether. Basically, `val useShortNames = declarations.find {it.element == firstEnumVariant} != null` ?